packages/docprovider/src/ydrive.ts

```diff
@@ -0,0 +1,234 @@
+// Copyright (c) Jupyter Development Team.
+// Distributed under the terms of the Modified BSD License.
+
+import {
+  DocumentChange,
+  IFactory,
+  ISharedDocument,
+  YDocument,
+  YFile,
+  YNotebook
+} from '@jupyter/ydoc';
+import { URLExt } from '@jupyterlab/coreutils';
+import { Contents, Drive, User } from '@jupyterlab/services';
+import { WebSocketProvider } from './yprovider';
+
+/**
+ * The url for the default drive service.
+ */
+const Y_DOCUMENT_PROVIDER_URL = 'api/yjs';
+
+/**
+ * A Collaborative implementation for an `IDrive`, talking to the
+ * server using the Jupyter REST API and a WebSocket connection.
+ */
+export class YDrive extends Drive {
+  /**
+   * Construct a new drive object.
+   *
+   * @param user - The user manager to add the identity to the awareness of documents.
+   */
+  constructor(user: User.IManager) {
+    super({ name: 'YDrive' });
+    this._user = user;
+    this._sharedPaths = new Set<string>();
+    this._providers = new Map<string, WebSocketProvider>();
+
+    this.sharedModelFactory = new SharedModelFactory(this._onCreate);
+  }
+
+  /**
+   * SharedModel factory for the YDrive.
+   */
+  readonly sharedModelFactory: IFactory;
+
+  /**
+   * Delete a file.
+   *
+   * @param localPath - The path to the file.
+   *
+   * @returns A promise which resolves when the file is deleted.
+   *
+   * #### Notes
+   * Uses the [Jupyter Notebook API](http://petstore.swagger.io/?url=https://raw.githubusercontent.com/jupyter/notebook/master/notebook/services/api/api.yaml#!/contents).
+   */
+  async delete(localPath: string): Promise<void> {
+    await super.delete(localPath);
+    // TODO should we remove the path from the sharedPath?
+    this._sharedPaths.delete(localPath);
```

Comment:
```suggestion
    // We are not removing the path from `sharedPaths` as multiple providers of the same file (with different model) may exist.
    // this._sharedPaths.delete(localPath);
```

There is something wrong with `_sharedPaths` because it does not deal with out of bound rename. We should track this somewhere for latter analysis.